src/lib/profiles/security/WeaveCert.cpp

```diff
@@ -894,6 +895,320 @@ NL_DLL_EXPORT uint32_t SecondsSinceEpochToPackedCertTime(uint32_t secondsSinceEp
+ *   format.
+ *
+ * @param  deviceId                Weave operational device Id.
+ * @param  devicePubKey            Weave operational device public key.
+ * @param  cert                    A pointer to a buffer where generated certificate to be written.
+ * @param  certBufSize             The length in bytes of the provided certificate buffer.
+ * @param  certLen                 The length in bytes of the generated certificate.
+ * @param  genCertSignature        A pointer to a function that generates ECDSA signature on the given
+ *                                 certificate hash using operational device private key.
+ *
+ * @retval #WEAVE_NO_ERROR         If Weave certificate was successfully generated.
+ */
+NL_DLL_EXPORT WEAVE_ERROR GenerateOperationalDeviceCert(uint64_t deviceId, EncodedECPublicKey& devicePubKey,
+                                                        uint8_t *cert, uint16_t certBufSize, uint16_t& certLen,
+                                                        GenerateECDSASignatureFunct genCertSignature)
+{
+    WEAVE_ERROR err;
+    TLVWriter writer;
+    TLVType containerType;
+    TLVType containerType2;
+    TLVType containerType3;
+    uint8_t *certDecodeBuf = NULL;
+    WeaveCertificateData *certData = NULL;
+
+    VerifyOrExit(genCertSignature != NULL, err = WEAVE_ERROR_INVALID_ARGUMENT);
+
+    VerifyOrExit(devicePubKey.ECPoint != NULL, err = WEAVE_ERROR_INVALID_ARGUMENT);
+
+    writer.Init(cert, certBufSize);
+
+    err = writer.StartContainer(ProfileTag(kWeaveProfile_Security, kTag_WeaveCertificate), kTLVType_Structure, containerType);
+    SuccessOrExit(err);
+
+    // Certificate serial number.
+    {
+        enum
+        {
+            kCertSerialNumber_Length          = 8,      // Length of the certificate serial number.
+            kCertSerialNumber_FirstByteMask   = 0x7F,   // Mask applied on the first byte of the key Id value.
+            kCertSerialNumber_FirstBytePrefix = 0x40,   // 4-bit Type value (0100) added at the beginning of the key Id value.
+        };
+        uint8_t certSerialNumber[kCertSerialNumber_Length];
+
+        // Generate a random value to be used as the serial number.
+        err = nl::Weave::Platform::Security::GetSecureRandomData(certSerialNumber, kCertSerialNumber_Length);
+        SuccessOrExit(err);
+
+        // Apply mask to avoid negative numbers.
+        certSerialNumber[0] &= kCertSerialNumber_FirstByteMask;
+
+        // Apply mask to guarantee the first byte is not zero.
+        certSerialNumber[0] |= kCertSerialNumber_FirstBytePrefix;
+
+        err = writer.PutBytes(ContextTag(kTag_SerialNumber), certSerialNumber, sizeof(certSerialNumber));
+        SuccessOrExit(err);
+    }
+
+    // Weave signature algorithm.
+    err = writer.Put(ContextTag(kTag_SignatureAlgorithm), static_cast<uint8_t>(kOID_SigAlgo_ECDSAWithSHA256 & ~kOIDCategory_Mask));
+    SuccessOrExit(err);
+
+    // Certificate issuer Id.
```

Comment:
Technically, this field is the certificate issuer _distinguished name_, or maybe DN for short.

Reply:
done